csrc/scheduler/matmul.cpp

```diff
@@ -51,6 +57,377 @@ void moveInnerBroadcastLeft(TensorView* tv, int number_of_inner_pos = 3) {
+     */
+
+    // row_stride in Z/nZ, where n is num_megabanks:
+    // assert(row_stride >= 0);
+    // assert(num_megabanks >= 0);
+    int row_stride_znz = row_stride % num_megabanks;
+
+    /* Consider the following function in Z/nZ:
+     *   f(i) = init + i * stride
+     * where init is the initial position of the pointer in the clock when we
+     * start the game, and stride is the number of ticks we move forward each
+     * time, and i is the number of times we move forward.
+     *
+     * In our problem, f(i) is the megabank of the `i`th row of the matrix, and
+     * `init` is the megabank of the 0th row of the matrix.
+     *
+     * One very important property of f(i) is:
+     * - if f(i1) == f(i2), then for every j, f(i1 + j) = f(i2 + j)
+     * This property is true because:
+     *   f(i1 + j) = f(i1) + j * stride = f(i2) + j * stride = f(i2 + j)
+     *
+     * The above property tells us, as we turn the clock forward:
+     * - initially, we will go to a never-visited tick in each turn, but,
+     * - at some point, we will return back to our original position, and,
+     * - after we return, we start repeat the pervious pattern again and again.
+     *
+     * As an example, consider f(i) where init = 0, stride = 6, under Z/8Z:
+     *     i  0 1 2 3 4 5 6 7
+     *   f(i) 0 6 4 2 0 6 4 2
+     * We can see that f(i) is repeating a pattern of four unique numbers
+     * "0 6 4 2" twice. In our bank conflict problem, this means we are using 4
+     * different megabanks, and we have a 2-way conflict.
+     *
+     * The question of interest is, does the above observation generalize? That
+     * is, does f(i) always repeat a pattern of p unique numbers q times? Note
+     * that p and q must satisfy p * q = n.
+     *
+     * The answer to the above question is: yes! Consider the following
+     * equation:
+     *    f(i1 + j) == f(i1)
+     * We want to know what is the smallest positive number j that makes the
+     * above equation true. Because this tells us in how many steps we will see
+     * repeat. This equation can be simplified as:
+     *   f(i1 + j) == f(i1) + j * stride == f(i1)
+     *   ==> j * stride == 0
+     *
+     * An important tool to study this equation is multiplicative inverse:
+     * https://en.wikipedia.org/wiki/Modular_multiplicative_inverse
+     * stride has an multiplicative inverse if and only if stride coprime with
+     * n, that is, g := gcd(stride, n) == 1. In such case, the solution to our
+     * equation j * stride == 0 is j = stride^(-1) * 0 = 0, that is: f(i) does
+     * not repeat, that is: there is no bank conflict.
+     */
+
+    int g = std::gcd(num_megabanks, row_stride_znz);
+    if (g == 1) {
+      return; // No need to swizzle in this case.
+    }
+
+    /* For the case where stride does not coprime with n, if n is a power of a
+     * prime, then we can write stride as:
+     *   stride = s * g
+     * where s coprime with n. Then the equation j * stride == 0 becomes:
```

Comment:
Not clear why `s` is coprime with `n`. 

```
stride = s * g
g = gcd of stride and n
```

So, let's say:
```
n = g * k
```

Then I can see `k` and `s` are coprime. Is `n` also coprime with `s`?

Reply:
There is a pre-condition `if n is a power of a prime`, so `n = p^k` then `g` must be `p^l` with `l <= k` (note that `0 <= stride < n` otherwise, we would be studying `stride % n` instead of `stride`). And `p` must not divide `s` because otherwise `g` would have been `p^(k+1)` instead of `p^k`. So `s` has no prime factor `p`, and `n` has no prime factor other than `p`, they coprime.

Reply:
Looks like this entire paragraph can be simplified. The proof I am using here is over complicated.

Solving `j * stride = 0` in `Z/nZ` is equivalent to solving `(j * stride) % n = 0` in `Z`, and according to 
https://github.com/NVIDIA/Fuser/blob/7a3b24ed43a9072fd771145828a4d38b34020589/csrc/expr_simplifier.h#L241
This is `(j * s) % k * g = 0` where `s = stride / g` and `k = n / g`.
This simplifies to `j * s = 0` in `Z/kZ`, because `s` coprime with `k`, then we must have `j = 0` in `Z/kZ`, that is, `j` is a multiple of `k` in `Z`, that is, the smallest positive `j` is `k`.

Reply:
> because otherwise `g` would have been `p^(k+1)` instead of `p^k`

Did you mean `p^(l+1)`?

Reply:
> > because otherwise `g` would have been `p^(k+1)` instead of `p^k`
> 
> Did you mean `p^(l+1)`?

Oh, yes, thanks for pointing out. I will update that reply.

Reply:
Updated this part in code